Revamp running programs with readout symmetrization

joshcombes opened the pull request:
Description
-----------
**Summary:** This PR revamps the existing readout symmetrization in `qc.run_symmetrized_readout`. 

`run_symmetrized_readout` is a method that runs a quil program in such a way that the readout error, or the confusion matrix, is made symmetric. Typically this involves running many variants of the original program. Enforcing symmetric readout error is useful in simplifying the assumptions in some near term error mitigation strategies, see e.g. `measure_observables`( which was originally developed for use in quantum chem applications and later for QCVV protocols).

**Details:** The previous implementation only symmetrized each single qubit confusion matrix. A swathe of internally collected experimental data shows that this kind of  symmetrization was insufficient for our purposes. Because of that a separate exhaustive symmetrization method was introduced to `measure_observables`.  The introduction of symmetrization to that code has made it more complicated than it needs to be.  Another big problem, specifically with exhaustive symmetrization, is that it comes with an exponential  (in the number of readout bits) overhead in additional experiments. E.g. for  `n` readout bits the minimum number of additional experiments would be `2^n-1` and in practice you would need `const * 2^n` additional experiments where the constant is of order 100 or 1000 depending on you use case.

In this revamp we introduce symmetrization methods that are statistically stronger than the previous implementation and two of the new methods have a linear scaling (in the number of readout bits) of the number of additional experiments. This PR has a number of advantages over doing the symmetrization in measure observables, the most important is it simplifies the logic in the measure observables.  
The symmetrization types available are:
- **exhaustive** uses every possible combination of flips (replicating `measure_observables` functionality)
- **trivial** that is no symmetrization (this is to aid testing)
- **orthogonal array (OA) with strength 1**  (replicating the old `run_symmetrized_readout` functionality)
- **OA with strength 2**
- **OA with strength 3**

By default a strength 3 OA is used; this ensures expectations of the form <b_k b_j b_i> --for any bits i,j,k-- will have symmetric readout errors. As a strength 3 OA is also a strength 2 and a strength 1 OA it also ensures <b_j b_i> and <b_i>  have symmetric readout errors for any bits b_j and b_i.

Use the notebook in this https://github.com/rigetti/pyquil/commit/66b1f2c957a25a07f3a60a9ec5350749dfe20b15 commit to explore the code. This commit has code that is a non standard way of adding noise models. I delete the notebook in subsequent commits.

Credit: a good chunk of the code in this PR is due to @kylegulshen e.g. `consolidate_symmetrization_outputs` see https://github.com/rigetti/forest-benchmarking/pull/133 . The original implementation / idea was @mpharrigan's;  it was symmetrization using an orthogonal array  with strength 1.

This should give you an idea of the cost of various readout symmetrization methods
![image](https://user-images.githubusercontent.com/2394563/58853606-3c0f2d00-86de-11e9-97d8-20889af5f0bf.png)

For larger numbers of qubits the linear scaling of the two and three designs becomes apparent
![image](https://user-images.githubusercontent.com/2394563/58857562-f2780f80-86e8-11e9-9d9e-d1371311c0bf.png)


**Related issues:**
- https://github.com/rigetti/qvm/issues/52 (comment: apparently the time overhead of a non parametric program can be reduced significantly)
- https://github.com/rigetti/pyquil/issues/843
- https://github.com/rigetti/pyquil/issues/844

Checklist
---------

- [x] The above description motivates these changes.
- [x] There is a unit test that covers these changes.
- [x] All new and existing tests pass locally and on Semaphore.
- [x] Parameters have type hints with [PEP 484 syntax](https://www.python.org/dev/peps/pep-0484/).
- [x] Functions and classes have useful sphinx-style docstrings.
- [x] (New Feature) this is a revamp of an old feature.
- [x] (Bugfix)  not a bug fix.

notmgsk commented:
My approval was accidental. Ignore it for now.

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -70,6 +73,245 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+    # valid_num_qubits = 4 * lambda - 1
+    valid_numbers = [4 * lam - 1 for lam in range(1, 70)]
+    # 4 * lambda
+    four_lam = int(min(x for x in valid_numbers if x >= num_qubits) + 1)
```
no need for int casting

joshcombes replied:
- [x] todo

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -70,6 +73,245 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+    four_lam = int(min(x for x in valid_numbers if x >= num_qubits) + 1)
+    H = hadamard(_next_power_of_2(four_lam))
+    # The minus sign in front of H fixes the 0 <-> 1 inversion relative to the reference [OATA]
+    design = ((-H[1:int(four_lam), 0:int(four_lam)] + 1) / 2).astype(int)
```
no need for int casting

joshcombes replied:
I removed the inner int casting. I am still doing it to the whole array.

- [x] todo

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -189,6 +431,81 @@ def run_symmetrized_readout(self, program: Program, trials: int) -> np.ndarray:
+        sym_programs, flip_arrays = symmetrization(program, meas_qubits, symm_type)
+        # This will be 1 or a power of two
+        num_sym_progs = len(sym_programs)
+        num_shots_per_prog = trials // num_sym_progs
```
I'm still confused by the number of trials here. You ensure that trials is even and greater than the number of meas_qubits, but consider 
```python
len(meas_qubits) = 6
trials = 10
symm_type = -1
num_sym_progs = 2 ** 6
num_shots_per_prog = trials // num_sym_progs = 10 / 2 ** 6 = 0
```

Is the check that trials is even supposed to be making trials the next power of two? Even so the example above would fail. 

Further wouldn't we want to try to optimize the number of shots here? So for ```len(measure_qubits) = 6``` and initial ```trials=8```:
1) with a one-design we  keep ```trials = 8``` and go with ```num_shots_per_prog = 5```
2) with a two-design we keep ```trials = 8``` and set ```num_shots_per_prog = 1```
3) for three-design we need ```trials = 16``` and ```num_shots_per_prog = 1``` 
4) for exhaustive we need ```trials = 2**6```

joshcombes replied:
Ok I just change the code to check if the minimum number of trials is achieve on a per symm_type basis. I also edited the docstrings to say that typically we want many more trials than the minimum.

Take a look now and let me know what you think.

joshcombes replied:
So the changes I am just about to push do the following. Given `progRAM = Program([I(x) for x in range(0,6)])`

```python
qc.run_symmetrized_readout_new(progRAM, trials=10, symm_type=-1)

UserWarning: Number of trials was too low, it is now 64
# Below here is printed out for debugging purposes
trials 64
sym progs 64
(floor division) num_shots_per_prog 1
```
and the output for the cases you give above is
(1) 
```python
trials 8
sym progs 2
(floor division) num_shots_per_prog 4
```
(2) 
```python
trials 8
sym progs 8
(floor division) num_shots_per_prog 1
```

(3) 
```
UserWarning: Number of trials was too low, it is now 16
trials 16
sym progs 16
(floor division) num_shots_per_prog 1
```

(4) 
```python
UserWarning: Number of trials was too low, it is now 64
trials 64
sym progs 64
(floor division) num_shots_per_prog 1
```

kylegulshen replied:
This sounds good to me. As you suggested in person it is probably good to warn the user if 
```num_shots_per_program * len(symm_progrs) < trials```
We should probably document well which number of trials for various options won't be changed.

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -70,6 +73,245 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+        if len(meas_qubits) == 0:
+            # corresponds to measuring identity; no program needs to be run.
+            results.append(np.array([[]]))
+            continue
```
We might want to remove this check since typically running a program without any measurements results in an empty bitstring along with a warning. 

In any case ```run_symmetrized_readout_new``` called with ```measure_qubits = 0``` won't trigger this condition since the input ```programs``` list will be empty.

joshcombes replied:
- [x] todo remove it

joshcombes commented on pyquil/api/_quantum_computer.py:
```diff
@@ -70,6 +73,245 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+    # arbitrary number of qubits are not known to exist.
+    num_expts, num_qubits = flip_matrix.shape
+    if len(meas_qubits) != num_qubits:
+        flip_matrix = flip_matrix[0:int(num_expts), 0:int(len(meas_qubits))]
```
- [x] todo

joshcombes commented on pyquil/api/_quantum_computer.py:
```diff
@@ -70,6 +73,245 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+
+
+def symmetrization(program: Program, meas_qubits: List[int], symm_type: int = 3) \
+        -> Tuple[List[Program], List[Tuple[bool]]]:
```
List[np.ndarray] right?

- [x] todo

joshcombes commented on pyquil/api/_quantum_computer.py:
```diff
@@ -70,6 +73,245 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+    for bitarray, flip_array in zip(outputs, flip_arrays):
+        if len(flip_array) == 0:
+            # happens when measuring identity.
+            # TODO: better way of handling identity measurement? (in _measure_bitstrings too)
```
- [x] todo

joshcombes commented on pyquil/api/_quantum_computer.py:
```diff
@@ -70,6 +73,245 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+           https://dx.doi.org/10.1007/978-1-4612-1478-6
+
+    :param num_qubits: minimum number of qubits the OA should run on.
+    :return: A numpy array representing the OQ with shape N by k
```
- [x] todo

joshcombes commented:
If the broad strokes of this PR is acceptable then what I propose to do next is

1. remove the old code `run_symmetrized_readout` as it is a special case of the new code where `symm_type = 1`.

2. rename `run_symmetrized_readout_new` to `run_symmetrized_readout` then check all the tests pass and include more tests related to the newly written code.

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -70,6 +73,273 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+        if trials < _next_power_of_2(2 * num_qubits):
+            trials = _next_power_of_2(2 * num_qubits)
+            warnings.warn('Number of trials was too low, it is now ' + str(trials))
+    return trials
```
```python
    min_num_trials = 2
    if symm_type == -1:
       min_num_trials = 2 ** num_qubits:
    elif symm_type == 2:
        def _f(x):
            return 4 * x - 1
        min_num_trials = min(_f(x) for x in range(1, 1024) if _f(x) >= num_qubits) + 1
    elif symm_type == 3:
       min_num_trials = _next_power_of_2(2 * num_qubits):

    if trials < min_num_trials:
        trials = min_num_trials
        warnings.warn('Number of trials was too low, it is now ' + str(trials))
    return trials
```

joshcombes replied:
- [x] todo

marcusps commented:
This looks absolutely awesome.  But one annoying terminology question: calling those arrays t-designs may lead to confusion (and I can't find sources where this is standard, but then again did not look so hard).  Can we call them t-orthogonal arrays (following the [Wikipedia convention](https://en.wikipedia.org/wiki/Orthogonal_array))? Or maybe t-strength orthogonal arrays (following [Neil Sloane's convention](http://neilsloane.com/oadir/index.html))?

joshcombes commented:
@marcusps thanks :)

I went through the code and docstrings. I think I have removed all instances of the word design. It is still in the plot above, but that was generated as a once off in a notebook.

In sentences I went for "orthogonal arrays with strength t".

mpharrigan commented:
very cool

msohaibalam commented:
This looks great to me as well! A few very broad things to consider before we merge this in:

- Why `-1` for exhaustive symmetrization? If the idea is to eventually retire this method, maybe this is an OK choice. But if not, perhaps we should consider a different number for this option.
- Assuming the answer to the above is no, one other thing on the list of things-to-do was to add a method for random sampling of bit flips for readout symmetrization (see https://github.com/rigetti/pyquil/issues/843). Doesn't have to be in this PR, but if we're going to add that, we (again) need to have a good numbering system for the various methods.
- Is there truly a need for separate methods for the various strengths, or can we consolidate the different functions `construct_strength_two_orthogonal_array` and `construct_strength_three_orthogonal_array` into a more generalizable `construct_strength_xx_orthogonal_array`? I'm more than likely missing some finer point here, but superficially at least, it seems that the use of orthogonal arrays is to identify a subset of the `2^n` possible bit-flipped programs that are carried out in the exhaustive symmetrization version.
- Finally, why commit out the example notebook? It was very helpful for me, and I'm sure it will be for others too! I'd advocate for committing that back in.
- Before the actual merge, there are some tiny tidying-up tasks that I can help point out (e.g. removing comments, minor typos in doc-strings, and maybe more importantly, changing function names e.g. incorporating `run_symmetrized_readout_new` to `run_symmetrized_readout`.
- Once this gets merged, we should incorporate all of this into `operator_estimation`. I can make specific tickets/issues for that once that occurs.

joshcombes commented:
Test failure is from `test_corrected_statistics_2q_nontrivial_nonentangled_state`, the issue has already been noted here https://github.com/rigetti/pyquil/issues/922

joshcombes commented:
Thanks and thanks for the feedback! 

I'll respond below point by point including the things we discussed in person to update those people following this thread.

> * Why `-1` for exhaustive symmetrization? If the idea is to eventually retire this method, maybe this is an OK choice. But if not, perhaps we should consider a different number for this option.

Good question. Basically the strength zero is `0`, strength one is `1`, strength two is `2` and strength three is `3`. It is possible that someone else might include strength four and so on. For that reason Kyle suggested -1 for exhaustive.

> * Assuming the answer to the above is no, one other thing on the list of things-to-do was to add a method for random sampling of bit flips for readout symmetrization (see #843). Doesn't have to be in this PR, but if we're going to add that, we (again) need to have a good numbering system for the various methods.

I'd love to see a Monte Carlo option and or a [Latin hypercube sampling](https://en.wikipedia.org/wiki/Latin_hypercube_sampling) option. I almost included them but stopped myself. Here are the problem I see. Take monte carlo as the example. The function would have to change a little because now you would have to specify a number say `symm_type=-2` for monte carlo and the number of random programs. So for $n$ bits that number could be anything from `0` to `2^n-1`. But the other symmetrization methods, e.g. the strength 2 OA, have a linear scaling with the number of additional experiments. Exhaustive has `2^n` so presumably the user will want some number less than that. So if it is close to either shouldn't they just use one of the other options ...

> * Is there truly a need for separate methods for the various strengths, or can we consolidate the different functions `construct_strength_two_orthogonal_array` and `construct_strength_three_orthogonal_array` into a more generalizable `construct_strength_xx_orthogonal_array`? I'm more than likely missing some finer point here, but superficially at least, it seems that the use of orthogonal arrays is to identify a subset of the `2^n` possible bit-flipped programs that are carried out in the exhaustive symmetrization version.

They are genuinely different methods for constructing orthogonal arrays. One is not a subset of the other. I would recommend keeping them separate.

> * Finally, why commit out the example notebook? It was very helpful for me, and I'm sure it will be for others too! I'd advocate for committing that back in.

Because the particular feature I was trying to illustrate (symmetrization of the two bit confusion matrix) only worked when I hardcoded the noise into the `api._quantum_computer` module. I just (temporarily) reintroduced it so you and others can play with it why reviewing.

If you can help me figure out a way that does not use this hack I can re include it. The basic problem is the noise models don't play well with run / compile. I could also just include the single bit confusion matrix although that could have been done with the old code alone and does not illustrate the power of this new code.


> * Before the actual merge, there are some tiny tidying-up tasks that I can help point out (e.g. removing comments, minor typos in doc-strings, and maybe more importantly, changing function names e.g. incorporating `run_symmetrized_readout_new` to `run_symmetrized_readout`.

In recent commits I have deleted the code in the old `run_symmetrized_readout` and replaced it with the code that was in   `run_symmetrized_readout_new`. As for all the other stuff, that would be awesome. Feel free to make commits directly to this branch.

> * Once this gets merged, we should incorporate all of this into `operator_estimation`. I can make specific tickets/issues for that once that occurs.

Agreed. We also spoke about the refactor that has happened in forest benchmarking ([see this file](https://github.com/rigetti/forest-benchmarking/blob/master/forest/benchmarking/observable_estimation.py)). Perhaps we can re-integrate that code back into pyQuil.

karalekas commented on requirements.txt:
```diff
@@ -4,6 +4,7 @@ antlr4-python3-runtime
 requests
 six
 networkx >= 2.0.0
+scipy
```
I'm not convinced we need to add this dependency just to tensor up some Hadamards

joshcombes replied:
- [x] todo

karalekas commented on pyquil/tests/test_quantum_computer.py:
```diff
@@ -178,6 +311,17 @@ def test_readout_symmetrization(forest):
+def test_run_symmetrized_readout_error():
+    # This test checks if the function runs for any possible input on a small number of qubits.
+    # Locally this test was run all 8 qubits, but it was slow.
+    qc = get_qc("8q-qvm")
```
let's use a fixture. if there's not already one for a QVM-backed QC object, we should make one.

joshcombes replied:
I'm pretty [green behind the ears](https://languagelog.ldc.upenn.edu/nll/?p=727) so I don't know what a fixture is or what it means to have a QVM-backed QC object, which means I don't understand what you are asking me to do. Can you help me with this one?

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -671,3 +695,340 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+        raise ValueError("symm_type must be one of the following ints [-1, 0, 1, 2, 3].")
+    # There is no need to test symm_type == 0 or symm_type == 1 as they require one and two
+    # trials respectively and that is ensured by this:
+    min_num_trials = 2
```
put in else block for clarity

joshcombes replied:
- [x] todo

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -154,40 +131,87 @@ def run(self, executable: Executable,
+        num_shots_per_prog = trials // len(sym_programs)
+
+        if num_shots_per_prog * len(sym_programs) < trials:
+            print("The number of trials was modified from " + str(trials) + " to " + str(
```
prefer f-strings, and log statements or warnings over print statements

joshcombes replied:
- [x] here todo
- [x] add f string below circa line 1080 todo

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -154,40 +131,87 @@ def run(self, executable: Executable,
+        :return: A numpy array of shape (trials, len(ro-register)) that contains 0s and 1s.
         """
-        Run a quil program in such a way that the readout error is made collectively symmetric
+        if type(symm_type) is not int:
```
prefer isinstance

joshcombes replied:
- [x] todo

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -154,40 +131,87 @@ def run(self, executable: Executable,
+
+        The symmetrization types are specified by an int; the types available are:
+        -1 -- exhaustive symmetrization uses every possible combination of flips
+         0 -- trivial that is no symmetrization
```
so are we saying that this should behave identically to `run`?

joshcombes replied:
Not like `run`,  however `run_symmetrized_readout` should behave like `run_and_measure`.

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -671,3 +695,340 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+        prog = program.copy()
+        ro = prog.declare('ro', 'BIT', len(meas_qubits))
+        for idx, q in enumerate(meas_qubits):
+            prog += MEASURE(q, ro[idx])
```
so for `run_symmetrized_readout` users are expected to pass a MEASURE-less Program, but for `run` the MEASUREs need to be there?

joshcombes replied:
Yes. `run_symmetrized_readout` appends gates to the end of their program and then appends MEASUREs.  We should be comparing `run_and_measure` and `run_symmetrized_readout`.

If you would like, we could strip MEASUREs and then append the gates and re-append MEASUREs. But this behavior would be unlike `run` and unlike `run_and_measure`.

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -154,40 +131,87 @@ def run(self, executable: Executable,
-        This means the probability of a bitstring ``b`` being mistaken for a bitstring ``c`` is
-        the same as the probability of ``not(b)`` being mistaken for ``not(c)``
+        if meas_qubits is None:
+            meas_qubits = list(program.get_qubits())
```
just double checking that this makes sense for the `None` case. if we pass a program with no MEASUREs and use symm_type = 0, then `run_symmetrized_readout` will append a DECLARE statement and MEASUREs, necessarily making it different from `run`

joshcombes replied:
Yes. It should be like `run_and_measure`.

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -671,3 +695,340 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+    return prog
+
+
+def _symmetrization(program: Program, meas_qubits: List[int], symm_type: int = 3) \
```
my gut says this code doesn't make sense in the QC module, but I need to think about where it belongs

joshcombes replied:
Agreed. I wasn't sure where to put it.

joshcombes commented:
@karalekas thanks for the feedback!

Let me know what else needs to be done to move forward with this PR.

joshcombes commented:
@karalekas what are your thoughts? I should say if you would like to make API or other changes feel free to do so on this branch :)

joshcombes commented:
Last failure is from the stochastic test `test_corrected_statistics_2q_nontrivial_nonentangled_state`
